cpp/src/arrow/compute/kernels/nth_to_indices.cc

```diff
@@ -0,0 +1,205 @@
+  std::shared_ptr<DataType> type_;
+
+ public:
+  /// \brief Kernel interface
+  ///
+  /// delegates to subclasses via NthToIndices()
+  virtual Status Call(FunctionContext* ctx, const Datum& values, uint64_t n,
+                      Datum* offsets) = 0;
+
+  /// \brief output type of this kernel
+  std::shared_ptr<DataType> out_type() const override { return uint64(); }
+
+  /// \brief single-array implementation
+  virtual Status NthToIndices(FunctionContext* ctx, const std::shared_ptr<Array>& values,
+                              uint64_t n, std::shared_ptr<Array>* offsets) = 0;
+
+  /// \brief factory for NthToIndicesKernel
+  ///
+  /// \param[in] value_type constructed NthToIndicesKernel will support
+  ///            values of this type
+  /// \param[out] out created kernel
+  static Status Make(const std::shared_ptr<DataType>& value_type,
+                     std::unique_ptr<NthToIndicesKernel>* out);
+};
+
+template <typename ArrayType>
+bool CompareValues(const ArrayType& array, uint64_t lhs, uint64_t rhs) {
+  return array.Value(lhs) < array.Value(rhs);
+}
+
+template <typename ArrayType>
+bool CompareViews(const ArrayType& array, uint64_t lhs, uint64_t rhs) {
+  return array.GetView(lhs) < array.GetView(rhs);
+}
+
+template <typename ArrowType, typename Comparator>
+class NthToIndicesKernelImpl : public NthToIndicesKernel {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+
+ public:
+  explicit NthToIndicesKernelImpl(Comparator compare) : compare_(compare) {}
+
+  Status NthToIndices(FunctionContext* ctx, const std::shared_ptr<Array>& values,
+                      uint64_t n, std::shared_ptr<Array>* offsets) {
+    return NthToIndicesImpl(ctx, std::static_pointer_cast<ArrayType>(values), n, offsets);
+  }
+
+  Status Call(FunctionContext* ctx, const Datum& values, uint64_t n, Datum* offsets) {
+    if (!values.is_array()) {
+      return Status::Invalid("NthToIndicesKernel expects array values");
+    }
+    auto values_array = values.make_array();
+    std::shared_ptr<Array> offsets_array;
+    RETURN_NOT_OK(this->NthToIndices(ctx, values_array, n, &offsets_array));
+    *offsets = offsets_array;
+    return Status::OK();
+  }
+
+  std::shared_ptr<DataType> out_type() const { return type_; }
+
+ private:
+  Comparator compare_;
+
+  Status NthToIndicesImpl(FunctionContext* ctx, const std::shared_ptr<ArrayType>& values,
+                          uint64_t n, std::shared_ptr<Array>* offsets) {
+    std::shared_ptr<Buffer> indices_buf;
+    int64_t buf_size = values->length() * sizeof(uint64_t);
+    RETURN_NOT_OK(AllocateBuffer(ctx->memory_pool(), buf_size, &indices_buf));
+
+    int64_t* indices_begin = reinterpret_cast<int64_t*>(indices_buf->mutable_data());
+    int64_t* indices_end = indices_begin + values->length();
+
+    std::iota(indices_begin, indices_end, 0);
+    *offsets = std::make_shared<UInt64Array>(values->length(), indices_buf);
+
+    if (n >= static_cast<uint64_t>(values->length())) {
```

Comment:
Shouldn't we have two cases here?
* `n` is exactly equal to the length: ok
* `n` strictly greater than the length: return `Status::Invalid`

Reply:
yes. done.